### ISSUE TITLE
for some reason the persist bookmark feature does not seem to work

### ISSUE THREAD

**Issue by i3d** [tristone13th/lspmark.nvim#3]
This is very helpful plugin btw, I love it. However, for some reason, when I exit neovim, and reopen again, all bookmarks I marked from the last session lost.

There is no error message of any sort, I am not too sure where I can debug from. It would be helpful if you could provide some guidance. 

What I did was simple, 

1. Use the toggle function to create a bookmark.
2. `Telescope lspmark` to verify that I can see this mark (which is yes)
3. exit neovim.
4. open neovim again.
5. Telescope lspmark` and I see an empty list, so basically the mark is gone.

**Comment by tristone13th** [comment]
The bookmarks are organized by cwd, you need to switch to your last cwf to load the bookmarks automatically.

**Comment by i3d** [comment]
I did try stay at the same dir opening the same file but I didn't see anything. What file do you store the records? I can try to see whether there is a such file exist... One thing to consider though, often the `cwd` we work on is within a d/cvs such git or hg, writing some state files under there typically is not very stable. would put the state file inside your plugin dir be workable?

**Comment by i3d** [comment]
ok, I did some debug, here is what I observed...

I made a mark for file `.vim/init.lua`. And here the plugin successfully created the `.json` file.

here is the content [elided the full path]

```
{"~/.vim/init.lua": {"12": {"_G.p": {"range": [40, 48, 0, 3], "marks": {"0": {"col": 0, "comment": "", "text": "function _G.p(...)"}}}}}}
```

file name is `%usr%local%xxx%home%xxx%.vim.json` [elided actual full path but format is here]

I reopen the same file `init.lua`. The marker is not loaded.

**Comment by i3d** [comment]
so nvm my previous previous comment, apparently your state file isn't written into the working dir.

**Comment by tristone13th** [comment]
Yes, it is saved at ~/.local/share/nvim/lspmark/.

Is this feature working now?

**Comment by karamanliev** [comment]
Great plugin, I like it a lot and I think it has lots of potential, thank you!

I experience the non persistent bookmarks problem as well. I see the file in `~/.local/share/nvim/lspmark/` that indeed persists my bookmarks, but for some reason when I `:q` and then re-enter nvim the marks ain't there. Here's the config I'm using:

```lua
  {
    'tristone13th/lspmark.nvim',
    keys = {
      {
        '<leader>bb',
        function()
          require('lspmark.bookmarks').toggle_bookmark({ with_comment = false })
        end,
        desc = 'Bookmark',
      },
      {
        '<leader>bB',
        function()
          require('lspmark.bookmarks').toggle_bookmark({ with_comment = true })
        end,
        desc = 'Bookmark (with comment)',
      },
      {
        '<leader>fb',
        '<cmd>Telescope lspmark<cr>',
        desc = '[F]ind [B]ookmarks',
      },
    },
    config = function()
      require('lspmark').setup()
      require('telescope').load_extension('lspmark')
    end,
  },
```

Edit: I tried with 3 different `cwd`s with 3 differnt `lsp`s and the result is the same - no persistance of the bookmarks.

**Comment by tristone13th** [comment]
Could you try to reproduce this issue following the sequence:

1. Use the toggle function to create a bookmark. run `pwd` to show the current directory, remember it as `project_dir`.
2. `Telescope lspmark` to verify that you can see this mark (which is yes)
3. exit neovim.
4. open neovim again.
5. `cd <project_dir>`.
6. `Telescope lspmark`

**Comment by i3d** [comment]
I did the steps and here is the screenshot.

the first screenshot is the result of step 1 and 2.

<img width="2892" alt="Screenshot 2024-06-23 at 14 25 25" src="https://github.com/tristone13th/lspmark.nvim/assets/105068/d79c7b43-25c5-444e-bb81-99193ecf6abd">

the second screenshot is the result after step 3.

<img width="2811" alt="Screenshot 2024-06-23 at 14 36 02" src="https://github.com/tristone13th/lspmark.nvim/assets/105068/3b905ffd-539b-48bc-974e-7d6547736878">

I edited the init.lua file under ~/.vim and that's the pwd and I didn't change any dir.

**Comment by 956MB** [comment]
I notice similar behavior when attempting to add bookmarks to files nested within a directory where one(s) already exist, or in a folder that is not the immediate parent of the file I'm in. Like maybe it has something to do with trying to add a bookmark to the same file `~/dotfiles/nvim/lua/plugins/db.lua` from different paths `~/dotfiles/nvim/lua/plugins` and `~/dotfiles`. Deleting all saved json files with the marks in `~/.local/share/nvim/lspmark`, and operating in `~/dotfiles/nvim/lua/plugins/db.lua`, I can see a few things happening:

1. `cd` to `~/dotfiles`.
2. Toggle bookmark.
3. `%Users%...%dotfiles.json` is created containing the single mark.
4. Quit, re-open nvim. Path is still `~/dotfiles`, and the bookmark in `db.lua` is gone, though it still remains in the json file.
5. `cd` to `~/dotfiles/nvim/lua/plugins`.
6. Toggle bookmark in `db.lua`, on a different line.
7. Quit, re-open nvim.
8. `%Users%...%dotfiles.json` is emptied "[]", and `%Users%...%dotfiles%nvim%lua%plugins.json` is created containing the single new bookmark.
9. `cd`'ing to the path where `db.lua` is (`~/dotfiles/nvim/lua/plugins`), adding/removing marks and quitting/re-opening nvim now persists the marks normally.

So it seems having multiple bookmarks on different lines in the same file, while having different `cwd` is either not intended behavior, or is just currently not working right.

**Comment by tristone13th** [comment]
Thanks for the detailed reproduction steps, however, I still cannot reproduce it. I'm using a C project maybe there are some differences between the C and the Lua language servers.

Could you show me your Lua config so I can try to find out what is really going wrong?

**Comment by 956MB** [comment]
Sure, it'd probably be easier if I just linked the whole thing and you could find what you might need from it. My [entire nvim config is here](https://github.com/956MB/dotfiles/tree/main/nvim).

I think it's possible my issue might be some tiny config specific thing, like maybe one or more things I've done incorrectly in my dotfiles, making it hard to find/reproduce. I'm not sure.

Edit: I will also try to mess around in a different project `not` in my dotfiles to see if I get the same issues.

**Comment by i3d** [comment]
Ok, so I did some debugging, one thing I noticed...

The json file is mangled with symbols. During loading, the display_bookmarks call is looking for [I put the debug error there]

```
Error detected while processing LspAttach Autocommands for "*":    
| Error executing lua callback: ...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:68: bookmark file does not exist: /Users/i3d/.vim/init.lua    
| stack traceback:    
|   [C]: in function 'error'    
|   ...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:68: in function 'display_bookmarks'    
|   ...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:511: in function 'lsp_calibrate_bookmarks'    
|   ...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:356: in function <...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:355>             
```

but the mangled file IS there...

```
ls $(realpath %Users%i3d%.vim.json )
 /Users/i3d/.local/share/nvim/lspmark/%Users%i3d%.vim.json
```

Could it because the code is trying to find the buffer file with normal name but the actually json bookmark file of that file name is mangled so they don't match anymore?

**Comment by tristone13th** [comment]
I don't think so, actually they are not the same file, one for storing the bookmarks information for the whole project (the mangled one) another is just a file inside the project (the not mangled one).

**Comment by i3d** [comment]
fyi, another suggestion might be that whether you could consider just having a single fixed name for storing the bookmarks and inside the bookmark json file, you could create structures to organize bookmarks that belong to different file target. I am not sure why the per-file solution is mangled, but if you don't necessarily (aka not a design limitation but just a choice of implementation) have to deal with, then I wonder may be to just maintain a single dedicated file , say 'bookmarks.json' and avoid all of per-file complexity all together would be something considerable. 

I think once the file is loaded and parsed to json structs, things will be fairly easy to navigate programmatically. I'd also assume that the content wouldn't get large enough that would start hurt performance... (fyi, if you could also perhaps install the load API as vim.schedule callback, then the plugin load time would be completely off the hook from the main startup, I suppose. The mark signs would then appear naturally when the load process is done).

**Comment by i3d** [comment]
so, a bit more debugging, I did another print before this line from telescope extension code

```
for file_name, kinds in pairs(bookmarks.bookmarks) do
```

and the result is ```{}```

**Comment by i3d** [comment]
ok, I think finally I know what happened...

here is the setup code

```
function M.setup()
  M.load_bookmarks() -- I did add this
  vim.api.nvim_create_autocmd({ "DirChanged" }, {
    callback = M.load_bookmarks,
    pattern = { "*" },
  })
  ...
end
```

I pull the ```M.load_bookmarks()``` explicitly out and called it. You know what, the marks appeared. :) So for whatever reason, that event perhaps didn't get trigger when open a file (that already had marks).

Another thing I found that are quite annoying is that marks associated with files can't be loaded by just opening that file from _any_ location. If the requirement is we have to go to the specific dir when we created the marks with the file, it would seem to be quite limited in usage (for example I have many many shell aliases in just invoking different files, e.g. v='nvim ~/.vim/init.lua', which supports me from any location)

**Comment by tristone13th** [comment]
> Another thing I found that are quite annoying is that marks associated with files can't be loaded by just opening that file from _any_ location. If the requirement is we have to go to the specific dir when we created the marks with the file, it would seem to be quite limited in usage (for example I have many many shell aliases in just invoking different files, e.g. v='nvim ~/.vim/init.lua', which supports me from any location)

Most of the time I use this plugin when developing in a project so I don't have the requirement to tweak single files, another issue is that the file-scope marks and project-scope marks may conflict so maintaining them will introduce complexity, anyway, if you have any better idea, don't hesitate to open a discussion or submit a PR!

**Comment by tristone13th** [comment]
@i3d @karamanliev @956MB 
I refresh the code a lot mainly for another issue #2 so I don't know if this issue still exist or not after the refreshing, would you mind having a test in your own configuration two see if this can be reproduced or not? Thanks:)

**Comment by karamanliev** [comment]
> @i3d @karamanliev @956MB I refresh the code a lot mainly for another issue #2 so I don't know if this issue still exist or not after the refreshing, would you mind having a test in your own configuration two see if this can be reproduced or not? Thanks:)

Sorry, I'm currently on a holiday without a computer around, will test when I'm back.

**Comment by karamanliev** [comment]
Ok, I'm back from vacation. Updated the plugin and tested, sadly I will report that it does not work as expected. I've recorded a video:

[Screencast from 2024-06-28 17-26-45.webm](https://github.com/tristone13th/lspmark.nvim/assets/22788016/6b6978ed-147a-45b8-830b-dc6777ceb274)

**Comment by tristone13th** [comment]
Hi @karamanliev , could you try running the

```
:lua require("lspmark.bookmarks").load_bookmarks()
```

after the second launch, then run the telescope to see if there are bookmarks listed there?

**Comment by karamanliev** [comment]
Doesn't work, for some reason telescope extension also stopped working after latest commits:

```error
Error executing Lua callback: Vim:E158: Invalid buffer name: 0
stack traceback:
	[C]: in function 'sign_getplaced'
	...l/share/nvim/lazy/lspmark.nvim/lua/lspmark/bookmarks.lua:144: in function 'lsp_calibrate_bookmarks'
	.../lazy/lspmark.nvim/lua/telescope/_extensions/lspmark.lua:17: in function <.../lazy/lspmark.nvim/lua/telescope/_extensions/lspmark.lua:13>
	...share/nvim/lazy/telescope.nvim/lua/telescope/command.lua:193: in function 'run_command'
	...share/nvim/lazy/telescope.nvim/lua/telescope/command.lua:259: in function 'load_command'
	...ocal/share/nvim/lazy/telescope.nvim/plugin/telescope.lua:108: in function <...ocal/share/nvim/lazy/telescope.nvim/plugin/telescope.lua:107>
```

**Comment by tristone13th** [comment]
There may be some breaking changes, please try to update and delete `~/.local/share/nvim/lspmark/` folder and retry.

**Comment by karamanliev** [comment]
> There may be some breaking changes, please try to update and delete `~/.local/share/nvim/lspmark/` folder and retry.

 I did already, doesn't help, sorry.

**Comment by tristone13th** [comment]
I just have a quick fix, could you try it now? Thanks.

**Comment by karamanliev** [comment]
yep, telescope extension works properly now.

**Comment by tristone13th** [comment]
Please try the latest commit and follow the latest README to try to reproduce this bug, I assume this has been fixed, feel free to reopen it if you have any concern.
